test/access_control.sh

```diff
@@ -132,85 +162,98 @@ function test_user_has_proper_access() {
-            os::log::info Checking access providing bogus token
-            test_message="invalid access from es with BOGUS token"
-            os::cmd::expect_success "run_curl_cmd 401 curl_es_pod_with_token $espod '/project.$proj.*/_count' BOGUS"
-
-            test_message="invalid access from kibana with BOGUS token"
-            os::cmd::expect_success "run_curl_cmd 403 curl_es_from_kibana $kpod $eshost '/project.$proj.*/_count' BOGUS"
-
-            # no token
-            os::log::info Checking access providing no username or token
-            test_message="invalid access from es with no token"
-            os::cmd::expect_success "run_curl_cmd 401 curl_es_pod_with_token $espod '/project.$proj.*/_count' ''"
-
-            test_message="invalid access from kibana with no token"
-            os::cmd::expect_success "run_curl_cmd 403 curl_es_from_kibana $kpod $eshost '/project.$proj.*/_count' ''"
-        fi
-    done
+    os::log::info See if user $user $negverb read alias /app
+    
+    os::log::info Checking access directly against ES pod...
+    test_message="$user $verb access /app index alias from es"
+    curl_es_pod $espod "/app/_count?pretty"
+    ## Access from elasticsearch pod will not go via es-proxy, so need to pass these headers
+    local xfuser='-H x-forwarded-user:'"$user"
+    local xocpns='-H x-ocp-ns:'"$user_project_list"
+    local xfroles='-H x-forwarded-roles:project_user'
+    local xff='-H x-forwarded-for:localhost'
+    os::cmd::expect_success_and_text \
+        "run_curl_cmd $expected_status curl_es_pod_with_token $espod '/app/_count' $test_token -XGET $xff $xfuser $xocpns $xfroles | get_count_from_json" \
+        "^$expected\$"
+
+    os::log::info Checking access from Kibana pod...
+    ## Access from outside elasticsearch pod, e.g. kiana pod, will go via es-proxy, so the headers will be added by es-proxy
+    test_message="$user $verb access /app index alias from kibana"
+    os::cmd::expect_success_and_text \
+        "run_curl_cmd $expected_status curl_es_from_kibana $kpod $eshost '/app/_count' $test_token -XGET | get_count_from_json" \
+        "^$expected\$"
+
+    # make sure no access with incorrect auth
+    # bogus token
+    os::log::info Checking access providing bogus token
+    test_message="invalid access from es with BOGUS token"
+    os::cmd::expect_success "run_curl_cmd 401 curl_es_pod_with_token $espod '/app/_count' BOGUS"
+
+    test_message="invalid access from kibana with BOGUS token"
+    os::cmd::expect_success "run_curl_cmd 500 curl_es_from_kibana $kpod $eshost '/app/_count' BOGUS"
+
+    # no token
+    os::log::info Checking access providing no username or token
+    test_message="invalid access from es with no token"
+    os::cmd::expect_success "run_curl_cmd 401 curl_es_pod_with_token $espod '/app/_count' ''"
+
+    test_message="invalid access from kibana with no token"
+    os::cmd::expect_success "run_curl_cmd 200 curl_es_from_kibana $kpod $eshost '/app/_count' ''"
```

Comment:
Any invalid response should give us something like a 401 or 403 I would think

Reply:
I wanted to discuss this. This testcase sends empty auth token, and from proxy logs i see that the user identified is kibana user. This is strange. This is why we get 200 rather than 40x.

```
time="2020-03-22T05:25:06Z" level=info msg="Serving request: &{GET /app/_count HTTP/1.1 1 1 map[User-Agent:[curl/7.29.0] Accept:[*/*] Testheader:[vimal] Authorization:[Bearer]] {} <nil> 0 [] false elasticsearch:9200 map[] map[] <nil> map[] 10.128.0.160:51642 /app/_count 0xc42047aa50 <nil> <nil> 0xc420449900}"
time="2020-03-22T05:25:06Z" level=info msg="Handling request \"authorization\""
time="2020-03-22T05:25:06Z" level=info msg="Request: &{GET /app/_count HTTP/1.1 1 1 map[User-Agent:[curl/7.29.0] X-Forwarded-User:[CN=system.logging.kibana,OU=OpenShift,O=Logging] Accept:[*/*] Testheader:[vimal] X-Forwarded-For:[localhost]] {} <nil> 0 [] false elasticsearch:9200 map[] map[] <nil> map[] 10.128.0.160:51642 /app/_count 0xc42047aa50 <nil> <nil> 0xc420449900}"
```

(I have modified proxy  to print complete incoming request, and added TestHeader from test case running locally to identify the request)

Reply:
It may be because it matches the certificates which are valid for the middleware server